drivers/rtc/rtc_rv8263.c

```diff
@@ -0,0 +1,780 @@
+int rv8263c8_calibration_set(const struct device *dev, int32_t calibration)
+{
+	int8_t offset;
+	int32_t test_mode0;
+	int32_t test_mode1;
+	int32_t offset_ppm_mode0;
+	int32_t offset_ppm_mode1;
+	const struct rv8263c8_config *config = dev->config;
+
+	/* NOTE: The RTC API is using a PPB (Parts Per Billion) value. The RTC is using PPM.
+	 * Here we calculate the offset when using MODE = 0.
+	 *  Formula from the application manual:
+	 *  Offset [ppm] = (calibration [ppb] / (4.34 [ppm] x 1000))
+	 */
+	offset_ppm_mode0 = calibration / 4340;
+
+	/* Here we calculate the offset when using MODE = 1.
+	 *  Formula from the application manual:
+	 *  Offset [ppm] = (calibration [ppb] / (4.069 [ppm] x 1000))
+	 */
+	offset_ppm_mode1 = calibration / 4069;
+
+	LOG_DBG("Offset Mode = 0: %i", offset_ppm_mode0);
+	LOG_DBG("Offset Mode = 1: %i", offset_ppm_mode1);
+
+	test_mode0 = offset_ppm_mode0 * 4340;
+	test_mode0 = calibration - test_mode0;
+	test_mode1 = offset_ppm_mode1 * 4069;
+	test_mode1 = calibration - test_mode1;
+
+	/* Compare the values and select the value with the smallest error. */
+	test_mode0 = test_mode0 < 0 ? -test_mode0 : test_mode0;
+	test_mode1 = test_mode1 < 0 ? -test_mode1 : test_mode1;
+	if (test_mode0 > test_mode1) {
+		LOG_DBG("Use fast mode (Mode = 1)");
+
+		/* Error with MODE = 1 is smaller -> Use MODE = 1. */
+		offset = RV8263_BM_FAST_MODE | (offset_ppm_mode1 & GENMASK(7, 0));
+	} else {
+		LOG_DBG("Use normal mode (Mode = 0)");
+
+		/* Error with MODE = 0 is smaller -> Use MODE = 0. */
+		offset = RV8263_BM_NORMAL_MODE | (offset_ppm_mode0 & GENMASK(7, 0));
+	}
+
+	LOG_DBG("Set offset value: %i", (offset & RV8263C8_BM_REGISTER_OFFSET));
+
+	return i2c_reg_write_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_OFFSET, offset);
+}
+
+int rv8263c8_calibration_get(const struct device *dev, int32_t *calibration)
+{
+	int err;
+	int32_t temp;
+	int8_t offset;
+	const struct rv8263c8_config *config = dev->config;
+
+	if (calibration == NULL) {
+		return -EINVAL;
+	}
+
+	err = i2c_reg_read_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_OFFSET, &offset);
+	if (err) {
+		return err;
+	}
+
+	/* Convert the signed 7 bit into a signed 8 bit value. */
+	if (offset & (0x01 << 6)) {
+		temp = offset | (0x01 << 7);
+	} else {
+		temp = offset & (0x3F);
+		temp &= ~(0x01 << 7);
+	}
+
+	LOG_DBG("Read offset: %i", temp);
+
+	if (offset & RV8263_BM_FAST_MODE) {
+		temp = temp * 4340L;
+	} else {
+		temp = temp * 4069L;
+	}
+
+	*calibration = temp;
+
+	return 0;
+}
+#endif
+
+static const struct rtc_driver_api rv8263c8_driver_api = {
+	.set_time = rv8263c8_time_set,
+	.get_time = rv8263c8_time_get,
+#if CONFIG_RTC_ALARM
+	.alarm_get_supported_fields = rv8263c8_alarm_get_supported_fields,
+	.alarm_set_time = rv8263c8_alarm_set_time,
+	.alarm_get_time = rv8263c8_alarm_get_time,
+	.alarm_is_pending = rv8263c8_alarm_is_pending,
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+	.alarm_set_callback = rv8263c8_alarm_set_callback,
+#endif
```

Comment:
If CONFIG_RTC_ALARM=y and int_gpios not OK, this driver can't pass test_alarm of  rtc_api. You could refer to /zephyr/tests/drivers/rtc/rtc_api/src/test_alarm.c .

Reply:
So shall I remove the check `DT_ANY_INST_HAS_PROP_STATUS_OKAY`?

Reply:
@bjarki-trackunit,
if the combination of enabled ALARM features without call-back is true, rtc.h and test_alarm.c/test_alarm_callback.c should be modified.
```
--- a/include/zephyr/drivers/rtc.h
+++ b/include/zephyr/drivers/rtc.h
@@ -389,7 +389,7 @@ static inline int z_impl_rtc_alarm_set_callback(const struct device *dev, uint16
        const struct rtc_driver_api *api = (const struct rtc_driver_api *)dev->api;

        if (api->alarm_set_callback == NULL) {
-               return -ENOSYS;
+               return -ENOTSUP;
        }

```

Reply:
I wouldn’t change the `rtc.h` because it returns the correct code (`ENOSYS`) for a driver not implementing this function. Instead, I would implement the function in the driver and return the correct code (`ENOTSUP`) to indicate that the functionality is not supported by the current configuration.

Reply:
Since there are only three existing drivers involved, I took the liberty to update them in this PR: #74248

Reply:
The rtc test suite should be updated to skip the test_alarm_callback if the callback can not be set :) Additionally, implementing the function and returning -ENOTSUP is the correct solution I think in case of the missing pin :)

Reply:
Ok I will add this check to the driver 👍

Reply:
Hi @cyliangtw @bjarki-trackunit I´ve added the required changes